README.md
# py_outlier_utils

[![ci-cd](https://github.com/UBC-MDS/py_outliers_utils/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-MDS/py_outliers_utils/actions/workflows/ci-cd.yml)

## Overview
As data rarely comes ready to be used and analyzed for machine learning right away, this package aims to help speed up the process of cleaning and doing initial exploratory data analysis specific to outliers. The package focuses on the tasks of identifying univariate outliers, providing summary of outliers like count, range of outliers, visualize them and giving functionality to remove them from data.

## Installation

```bash
$ pip install py_outlier_utils
```

## Functions
The three functions contained in this package are as follows:

- `outlier_identifier`: A function to identify outliers in the dataset and provide their summary as an output
- `visualize_outliers`: A function to generate the eda plots highlighting outliers providing additional functionality to visualize them
- `trim_outliers`: A function to generate outlier free dataset by imputing them with mean, median or trim entire row with outlier from dataset.

## Our Place in the Python Ecosystem
While Python packages with similar functionalities exist, this package aims to provide summary, visualization of outliers in a single package with an additional functionality to generate outlier-free dataset. Few packages with similar functionality are as follows:

- [pyod](https://pypi.org/project/pyod/)
- [python-outlier](https://pypi.org/project/python-outlier/)

## Usage

`py_outlier_utils` can be used to deal with the outliers in a dataset and plot the distribution of the dataset. More information can be found in [Example usage](https://py-outliers-utils.readthedocs.io/en/latest/example.html).

## Contributing

This package is authored by Karanpreet Kaur, Linhan Cai, Qingqing Song as part of the course project in DSCI-524 (UBC-MDS program). You can see the list of all contributors in the contributors tab.

We welcome and recognize all contributions. If you wish to participate, please review our [Contributing guidelines](CONTRIBUTING.md)

## License

`py_outlier_utils` is licensed under the terms of the MIT license.

## Credits

`py_outlier_utils` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.2.9 (2022-01-29)


## v1.2.8 (2022-01-29)


## v1.2.7 (2022-01-29)


## v1.2.6 (2022-01-28)


## v1.2.5 (2022-01-28)


## v1.2.4 (2022-01-28)


## v1.2.3 (2022-01-28)
### Documentation
* Updated example ([`c4a6da7`](https://github.com/UBC-MDS/py_outliers_utils/commit/c4a6da750b25c3222551f0e091898a5aa274a53f))

## v0.1.1 (2022-01-27)
### Documentation
* Updated example ([`c4a6da7`](https://github.com/UBC-MDS/py_outliers_utils/commit/c4a6da750b25c3222551f0e091898a5aa274a53f))

## v0.1.0 (11/01/2022)

- First release of `outliers`!

tests/test_outliers.py
import pandas as pd
from pandas._testing import assert_frame_equal
import numpy as np
import pytest
from pytest import raises
from py_outliers_utils.outliers import outlier_identifier

# Arrange
@pytest.fixture
def input_iqr():
    df = pd.DataFrame({ 'SepalLengthCm' : [5.2, 4.9, 4.7, 5.5, 5.1, 5.0, 6.0, 54, 50, 5.3, 5.1],
                        'SepalWidthCm' :  [-5.4, 1.4, -5.7, 0.2, 0.7, 1.6, 1.2, 1.4, 1.8, 1.5, 2.1],
                        'PetalWidthCm' :  [-40, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5],
                        'class': ['Iris Setosa', 'Iris Versicolour', 'Iris Virginica', 'Iris Setosa', 'Iris Versicolour', 'Iris Virginica', 'Iris Virginica', 
                                'Iris Setosa', 'Iris Versicolour', 'Iris Setosa', 'Iris Versicolour']
})
    return df

# Arrange
@pytest.fixture
def input_zscore():
    df = pd.DataFrame({ 'SepalLengthCm' : [5.1, 9, 6.5, 5.5, 6.7, 10.4, -54, 15, 5.2, 5.3, 5.1, 5.1, 4.9, 5.6, 6.5, 5.5, 6.7, 10.4, -55, 15, 5.2, 5.3, 5.1],
                        'SepalWidthCm' : [5.1, 9, 6.5, 5.5, 6.7, 10.4, 68, 15, 5.2, 5.3, 5.1, 5.1, 4.9, 5.6, 6.5, 5.5, 6.7, 10.4, 70, 15, 5.2, 5.3, 5.1],
                        'PetalWidthCm' :  [-40, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5, 6, 7, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5],
                        'RootLengthCm': [60, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5, 6, 7, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5],
                        'class': ['Iris Setosa', 'Iris Versicolour', 'Iris Virginica', 'Iris Setosa', 'Iris Versicolour', 'Iris Virginica', 'Iris Virginica', 
                                  'Iris Setosa', 'Iris Versicolour', 'Iris Setosa', 'Iris Versicolour', 'Iris Setosa', 'Iris Versicolour', 'Iris Virginica', 
                                  'Iris Setosa', 'Iris Versicolour', 'Iris Virginica', 'Iris Virginica', 'Iris Setosa', 'Iris Versicolour', 'Iris Setosa', 
                                  'Iris Versicolour', 'Iris Setosa']
                   })
    return df


# Arrange
@pytest.fixture
def summary_Zscore():
    compare_df = pd.DataFrame({'SepalLengthCm' : [2, '8.7%', 1.74, 5.5, 17.99, (-55.0, -54.0), np.nan],
                               'SepalWidthCm' :  [2, '8.7%', 12.48, 5.6, 18.08, np.nan, (68.0, 70.0)],
                               'PetalWidthCm' :  [1, '4.35%', -0.47, 0.4, 8.87, -40.0, np.nan],
                               'RootLengthCm' :  [1, '4.35%', 3.88, 0.4, 12.42, np.nan, 60.0]}, 
                               index=['outlier_count', 'outlier_percentage', 'mean', 'median', 'std', 'lower_range', 'upper_range'])
    return compare_df

# Arrange
@pytest.fixture
def summary_IQR():
    compare_df_iqr = pd.DataFrame({'SepalLengthCm' : [2, '18.18%', 13.71, 5.2, 18.96, np.nan, (50.0, 54.0)],
                                   'SepalWidthCm' :  [2, '18.18%', 0.07, 1.4, 2.83, (-5.7, -5.4), np.nan],
                                   'PetalWidthCm' :  [2, '18.18%', -2.9, 0.3, 12.38, -40.0, 5.0]}, 
                               index=['outlier_count', 'outlier_percentage', 'mean', 'median', 'std', 'lower_range', 'upper_range'])
    return compare_df_iqr

# Arrange
@pytest.fixture
def outlier_Zscore():
    compare_outlier_df = pd.DataFrame({'SepalLengthCm' : [5.1, 9, 6.5, 5.5, 6.7, 10.4, -54, 15, 5.2, 5.3, 5.1, 5.1, 4.9, 5.6, 6.5, 5.5, 6.7, 10.4, -55, 15, 5.2, 5.3, 5.1],
                                       'SepalWidthCm' : [5.1, 9, 6.5, 5.5, 6.7, 10.4, 68, 15, 5.2, 5.3, 5.1, 5.1, 4.9, 5.6, 6.5, 5.5, 6.7, 10.4, 70, 15, 5.2, 5.3, 5.1],
                                       'PetalWidthCm' :  [-40, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5, 6, 7, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5],
                                       'RootLengthCm': [60, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5, 6, 7, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5],
                                       'outlier': [True, False, False, False, False, False, True, False, False, False, False, False, False, False, False, False, False, False, True, False, False, False, False]
                                  })
    return compare_outlier_df

# Arrange
@pytest.fixture
def outlier_IQR():
    compare_outlier_IQR_df = pd.DataFrame({'SepalLengthCm' : [5.2, 4.9, 4.7, 5.5, 5.1, 5.0, 6.0, 54, 50, 5.3, 5.1],
                                           'SepalWidthCm' :  [-5.4, 1.4, -5.7, 0.2, 0.7, 1.6, 1.2, 1.4, 1.8, 1.5, 2.1],
                                           'PetalWidthCm' :  [-40, 0.2, 0.2, 0.3, 0.2, 0.5, 0.5, 0.6, 0.4, 0.2, 5], 
                                           'outlier': [True, False, True, False, False, False, False, True, True, False, True]
                                  })  
    return compare_outlier_IQR_df               

def test_outlier_identifier(input_iqr, input_zscore, summary_Zscore, summary_IQR, outlier_Zscore, outlier_IQR):
    """
    Test existing functionalities of outlier_identifier(), which include
    returning a dataframe in the right format with input arguments as columns, identifier and return_df argument.
    6 tests in total.
    """

   # Tests whether data is not of dataframe raises TypeError
    with raises(TypeError, match=r"passed dataframe is of type list, should be DataFrame"):
        outlier_identifier([4, None, 4, 7])

   # Tests whether columns passed in incorrect type raises TypeError
    with raises(TypeError, match=r"passed columns is of type set, should be list or NoneType"):
        outlier_identifier(input_iqr, columns= {'SepalLengthCm'})

    # Tests whether return_df passed in incorrect type raises TypeError
    with raises(TypeError, match=r"passed return_df is of type str, should be bool with value as True or False"):
        outlier_identifier(input_iqr, return_df='True')

    # Tests whether if columns are None or type list
    with raises(TypeError, match=r"passed columns is of type set, should be list or NoneType"):
        outlier_identifier(input_iqr, columns={'SepalLengthCm'})

    # Tests whether if identifier is of type str
    with raises(TypeError, match=r"passed identifier is of type list, should be string with value 'Z_score' or 'IQR'"):
        outlier_identifier(input_iqr, columns=['SepalLengthCm'], identifier=['Z_score'])

    # Tests whether wrong identifier passed raises ValueError
    with raises(ValueError):
        outlier_identifier(input_iqr, columns= ['SepalLengthCm'], identifier='both')

    # Tests whether empty dataframe or dataframe with all NAN raises ValueError
    with raises(ValueError):
        outlier_identifier(pd.DataFrame(), columns= ['SepalLengthCm'])

    


    ## Unit test cases

    # Test if output with identifier = 'Z_score' and return_df = False (default) matches with expected output summary_Zscore 
    # (checks if condition - 1)
    assert_frame_equal(outlier_identifier(input_zscore, identifier='Z_score'), summary_Zscore), "The returned dataframe using outlier_identifier is not correct"

    # Test if output with identifier = 'Z_score' and return_df = True matches with expected output outlier_Zscore 
    # (checks if condition - 2)
    assert_frame_equal(outlier_identifier(input_zscore, identifier='Z_score', return_df=True), outlier_Zscore), "The returned dataframe using outlier_identifier is not correct"

    # Test if output with identifier = 'IQR' (default) and return_df = False (default) matches with expected output summary_IQR 
    # (checks if condition - 3)
    assert_frame_equal(outlier_identifier(input_iqr, identifier='IQR'), summary_IQR), "The returned dataframe using outlier_identifier is not correct"

    # Test if output with identifier = 'IQR' (default) and return_df = True matches with expected output outlier_IQR 
    # (checks if condition - 4)
    assert_frame_equal(outlier_identifier(input_iqr, identifier='IQR', return_df=True), outlier_IQR), "The returned dataframe using outlier_identifier is not correct"
    
    # Test if output has all numeric columns from the list of columns passed 
    assert outlier_identifier(input_iqr, columns=['SepalLengthCm', 'SepalWidthCm', 'class']).columns.tolist() == ['SepalLengthCm', 'SepalWidthCm'], "The columns returned in the dataframe is not correct"


tests/test_trim_outliers.py
import pandas as pd
from pandas._testing import assert_frame_equal
import numpy as np
import pytest
from pytest import raises
from py_outliers_utils.trim_outliers import trim_outliers


def test_trim_outliers():
    """
    A function to test whether the output of trim_outliers is correct.

    """
    test_df = pd.DataFrame(
        {
            "SepalLengthCm": [5.1, 4.9, 4.7, 5.5, 5.1, 50, 5.4, 5.0, 5.2, 5.3, 5.1],
            "SepalWidthCm": [1.4, 1.4, 20, 2.0, 0.7, 1.6, 1.2, 1.4, 1.8, 1.5, 2.1],
            "PetalWidthCm": [0.2, 0.2, 0.2, 0.3, 0.4, 0.5, 0.5, 0.6, 0.4, 0.2, 5],
            "Species": [
                "Iris-setosa",
                "Iris-virginica",
                "Iris-germanica",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
            ],
        }
    )

    test_column = ["SepalLengthCm", "SepalWidthCm", "PetalWidthCm"]

    median_output = pd.DataFrame(
        {
            "SepalLengthCm": [5.1, 4.9, 4.7, 5.5, 5.1, 5.1, 5.4, 5.0, 5.2, 5.3, 5.1],
            "SepalWidthCm": [1.4, 1.4, 1.5, 2.0, 0.7, 1.6, 1.2, 1.4, 1.8, 1.5, 2.1],
            "PetalWidthCm": [0.2, 0.2, 0.2, 0.3, 0.4, 0.5, 0.5, 0.6, 0.4, 0.2, 0.4],
            "Species": [
                "Iris-setosa",
                "Iris-virginica",
                "Iris-germanica",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
            ],
        }
    )

    trim_output = pd.DataFrame(
        {
            "SepalLengthCm": [5.1, 4.9, 5.5, 5.1, 5.4, 5.0, 5.2, 5.3],
            "SepalWidthCm": [1.4, 1.4, 2.0, 0.7, 1.2, 1.4, 1.8, 1.5],
            "PetalWidthCm": [0.2, 0.2, 0.3, 0.4, 0.5, 0.6, 0.4, 0.2],
            "Species": [
                "Iris-setosa",
                "Iris-virginica",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
            ],
        }
    )

    mean_output = pd.DataFrame(
        {
            "SepalLengthCm": [5.1, 4.9, 4.7, 5.5, 5.1, 9.21, 5.4, 5.0, 5.2, 5.3, 5.1],
            "SepalWidthCm": [1.4, 1.4, 3.19, 2.0, 0.7, 1.6, 1.2, 1.4, 1.8, 1.5, 2.1],
            "PetalWidthCm": [0.2, 0.2, 0.2, 0.3, 0.4, 0.5, 0.5, 0.6, 0.4, 0.2, 0.77],
            "Species": [
                "Iris-setosa",
                "Iris-virginica",
                "Iris-germanica",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
            ],
        }
    )

    column_output = pd.DataFrame(
        {
            "SepalLengthCm": [5.1, 4.9, 4.7, 5.5, 5.1, 9.21, 5.4, 5.0, 5.2, 5.3, 5.1],
            "SepalWidthCm": [1.4, 1.4, 20, 2.0, 0.7, 1.6, 1.2, 1.4, 1.8, 1.5, 2.1],
            "PetalWidthCm": [0.2, 0.2, 0.2, 0.3, 0.4, 0.5, 0.5, 0.6, 0.4, 0.2, 5],
            "Species": [
                "Iris-setosa",
                "Iris-virginica",
                "Iris-germanica",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
                "Iris-setosa",
            ],
        }
    )

    numeric_only_df = pd.DataFrame({
        'SepalLengthCm' : [5.1, 4.9, 4.7, 5.5, 5.1, 50, 5.4, 5.0, 5.2, 5.3, 5.1],
        'SepalWidthCm' : [1.4, 1.4, 20, 2.0, 0.7, 1.6, 1.2, 1.4, 1.8, 1.5, 2.1],
        'PetalWidthCm' : [0.2, 0.2, 0.2, 0.3, 0.4, 0.5, 0.5, 0.6, 0.4, 0.2, 5]
    })

    numeric_only_out = pd.DataFrame({
        'SepalLengthCm': [5.1, 4.9, 5.5, 5.1, 5.4, 5.0, 5.2, 5.3],
        'SepalWidthCm': [1.4, 1.4, 2.0, 0.7, 1.2, 1.4, 1.8, 1.5],
        'PetalWidthCm' :[0.2, 0.2, 0.3, 0.4, 0.5, 0.6, 0.4, 0.2],
    })
    # Test if the imput is not dataFrame
    with raises(TypeError):
        trim_outliers("not dataframe")

    # Test if columns input is not list
    with raises(TypeError):
        trim_outliers(test_df, columns=2)

    # Test if input column list is in the dataframe
    with raises(Exception):
        trim_outliers(test_df, columns=["not in"])

    # Test if method input is not one of three methods provided
    with raises(Exception):
        trim_outliers(test_df, columns=["SepalLengthCm"], method="no")
        
    assert pd.DataFrame.equals(
        trim_outliers(test_df, test_column, method="mean"),
        mean_output,
    ), "The mean method is not correct"     

    assert pd.DataFrame.equals(
        trim_outliers(test_df, test_column, method="median"),
        median_output,
    ), "The median method is not correct"
    
    assert pd.DataFrame.equals(
        trim_outliers(
            test_df, columns=["SepalLengthCm"], method="mean"
        ),
        column_output,
    ), "The selected column method is not correct"


tests/test_visualize_outliers.py
from cmath import nan
from py_outliers_utils.visualize_outliers import visualize_outliers
import pandas as pd
import numpy as np
from pytest import raises
import altair as alt

def df():

    df = pd.DataFrame({"a":np.random.normal(100, 30, 100),
                  "b":np.random.normal(8, 5, size=100),
                  "c":np.random.randint(100, size=100),
                 "d":["A","B", "C", "D"] * 25})

    return df

def test_output_columns():

    ## Test the used columns of function result
    assert visualize_outliers(df(), columns = ['a']).data.variable.unique() == ['a'], 'Given columns should be mapped to the facet.'

def test_output_chart():

    ## Test the type of function result
    output_t = visualize_outliers(df(), columns = ['a', 'b', 'd'], type='violin')
    assert isinstance(
        output_t, alt.vegalite.v4.api.FacetChart
    ), "Altair Facet Chart object should be returned."

def test_output_type_vio():

    ## Test the type of function result
    assert visualize_outliers(df(), type='violin').columns == 5, 'Return chart is not of the given type.'

def test_output_type_box():

    ## Test the type of function result
    assert visualize_outliers(df(), type='boxplot').columns == 4, 'Return chart is not of the given type.'

def test_invalid_df():
    ## Test the exceptions of invalid input
    with raises(TypeError):
        visualize_outliers("572")

def test_invalid_df_non():
    ## Test the exceptions of invalid input
    dfn = pd.DataFrame({"a":[nan, nan],
    "b":[1,2]})
    with raises(ValueError):
        visualize_outliers(dfn)

def test_invalid_col():
    with raises(TypeError):
        visualize_outliers(df(), columns = "gsgs")


def test_invalid_type():
    with raises(ValueError):
        visualize_outliers(df(), type='satosugu')
        

src/outliers/visualize_outliers.py
import pandas as pd
import altair as alt

def visualize_outliers(dataframe, columns=None, type='violin'):
    """
    A function that plot the distribution of the given data.
    
    Parameters
    ----------
    dataframe : pandas.core.frame.DataFrame
        The target dataframe where the function is performed.
    
    columns : list, default=None
        The target columns where the function needed to be performed. Default is None, the function will check all columns.
    
    type : string, default='violin'
        The method of plotting the distribution.
        - if "violin" : Return a violin plot with boxplot layer
        - if "boxplot" : Return a boxplot


    Returns
    -------
    altair.vegalite.v4.api.Chart
        an altair plot with data distribution.
    
    Examples
    --------
    >>> import pandas as pd
    >>> import altair as alt
        
    >>> df = pd.DataFrame({
    >>>    'SepalLengthCm' : [0.1, 4.9, 52.7, 5.5, 5.1, 50, 5.4, 179.0, 5.2, 5.3, 5.1],
    >>>    'SepalWidthCm' : [1.4, 1.4, 20, 2.0, 0.7, 1.6, 1.2, 14, 1.8, 1.5, 2.1],
    >>>    'PetalWidthCm' : [0.2, 0.2, 0.2, 0.3, 0.4, 0.5, 0.5, 0.6, 0.4, 0.2, 5]
    >>> })

    >>> visualize_outliers(df, columns=['SepalLengthCm', 'SepalWidthCm'])
    """

    ## Handle dataframe type error (Check if dataframe is of type Pandas DataFrame)
    if not isinstance(dataframe, pd.DataFrame):
        raise TypeError(f"passed dataframe is of type {type(dataframe).__name__}, should be DataFrame")
        
    ## Handle empty dataframe or dataframe with all NAN
    if dataframe.empty or dataframe.dropna().empty:
        raise ValueError("passed dataframe is None")
    
    ## Handle columns type error (Check if columns are None or type list)
    if not columns is None and not isinstance(columns, list):
        raise TypeError(f"passed columns is of type {type(columns).__name__}, should be list or NoneType")
    
    ## Handle type Value error (Check if type has value 'violin' or 'boxplot')
    if type!='violin' and type!='boxplot':
        raise ValueError("passed type should have value 'violin' or 'boxplot'")
    
    ## Select given columns
    if columns is None:
        df = dataframe
    else:
        df = dataframe[columns]
    
    ## Select numeric columns
    dfnumeric = df._get_numeric_data() 
    
    ## Melt dataframe for plot
    dfmelt = dfnumeric.melt()  
    
    ## Plot according type
    if type == 'violin':
        
        ## The boxplot layer
        boxplot = alt.Chart().mark_boxplot(color='black').encode(
                            alt.Y('value')
                    ).properties(width=100)
        
        ## The violinplot layer
        violin = alt.Chart().transform_density(
                'value',
                as_=['value', 'density'],
                groupby=['variable']
            ).mark_area(orient='horizontal').encode(
                y=alt.Y('value'),
                color=alt.Color('variable:N', legend=None),
                x=alt.X(
                    'density:Q',
                    stack='center',
                    impute=None,
                    title=None,
                    scale=alt.Scale(nice=False,zero=False),
                    axis=alt.Axis(labels=False, values=[0], grid=False, ticks=True),
                ),
            )
        
        ## Overlay two layers
        violinbox = alt.layer(violin, boxplot, data=dfmelt).facet('variable:N', columns=5).resolve_scale(x=alt.ResolveMode("independent"))
        return violinbox
    
    elif type == 'boxplot':
        boxplot = alt.Chart(dfmelt).mark_boxplot().encode(
                    alt.X(title = None),
                    alt.Y("value"),
                    alt.Color("variable", legend = None) 
                ).properties(
                    width=100
                ).facet('variable:N', columns=5)
        return boxplot


src/py_outlier_utils/trim_outliers.py
import pandas as pd
import numpy as np
from py_outlier_utils.outliers import outlier_identifier

def trim_outliers(dataframe, columns=None, identifier='IQR', method='trim'):
    """
    A function to generate outlier free dataset by imputing them with mean, median or trim entire row with outlier from dataset.
     ----------
    dataframe : pandas.core.frame.DataFrame
        The target dataframe where the function is performed.
    columns : list, default=None
        The target columns where the function needed to be performed. Default is None, the function will check all columns
    identifier : string
        The method of identifying outliers.
        - if "Z_score" : Use z-test with threshold of 3
        - if "IQR" : Use IQR (Inter Quantile range) to identify outliers
    method : string
        The method of dealing with outliers.
            - if "trim" :  remove completely rows with data points having outliers.
            - if "median" : replace outliers with median values
            - if "mean" : replace outliers with mean values
            
    Return
    -------
    pandas.core.frame.DataFrame
        a dataframe which the outlier has already process by the chosen method.
        
    Examples
    --------
    >>> import pandas as pd
    
    >>> df = pd.DataFrame({
    >>>    'SepalLengthCm' : [5.1, 4.9, 4.7, 5.5, 5.1, 50, 5.4, 5.0, 5.2, 5.3, 5.1],
    >>>    'SepalWidthCm' : [1.4, 1.4, 20, 2.0, 0.7, 1.6, 1.2, 1.4, 1.8, 1.5, 2.1],
    >>>    'PetalWidthCm' : [0.2, 0.2, 0.2, 0.3, 0.4, 0.5, 0.5, 0.6, 0.4, 0.2, 5]
    >>> })
    >>> trim_outliers(df, columns=['SepalLengthCm', 'SepalWidthCm', 'PetalWidthCm'],identifier='Z_score', method='trim')
    	 SepalLengthCm  	SepalWidthCm	   PetalWidthCm
    0	5.1	                1.4	                0.2
    1	4.9	                1.4	                0.2
    2	5.5	                2.0	                0.3
    3	5.1	                0.7	                0.4
    4	5.4	                1.2             	0.5
    5	5.0	                1.4	                0.6
    6	5.2	                1.8	                0.4
    7	5.3	                1.5	                0.2
    """

    # Handle dataframe type error (Check if dataframe is of type Pandas DataFrame)
    if not isinstance(dataframe, pd.DataFrame):
        raise TypeError(f"passed dataframe is of type {type(dataframe).__name__}, should be DataFrame")
        
    # Handle empty dataframe or dataframe with all NAN
    if dataframe.empty or dataframe.dropna().empty:
        raise ValueError("passed dataframe is None")
    
    # Handle columns type error (Check if columns are None or type list)
    if not columns is None and not isinstance(columns, list):
        raise TypeError(f"passed columns is of type {type(columns).__name__}, should be list or NoneType")
    
    # Handle identifier type error (Check if identifier is of type str)
    if not isinstance(identifier, str):
        raise TypeError(f"passed identifier is of type {type(identifier).__name__}, should be string with value 'Z_score' or 'IQR'")
    
    # Handle identifier Value error (Check if identifier has value 'Z_score' or 'IQR')
    if identifier!='Z_score' and identifier!='IQR':
        raise ValueError("passed identifier should have value 'Z_score' or 'IQR'")
        
    # Handle method type error (Check if identifier method is 'trim', 'median' or 'mean')
    if method not in ("trim", "median", "mean"):
        raise Exception("The method must be -trim- or -median- or -mean-")


    dataframe_copy = dataframe.copy()
    
    if columns is None:
        columns = list(dataframe.columns)
    numeric_columns = dataframe[columns].select_dtypes('number').columns
    df = outlier_identifier(dataframe, columns=columns, identifier=identifier, return_df=True)
    if method == 'trim':
        outlier_index = df[df['outlier']==True].index
        dataframe_copy = dataframe_copy.drop(outlier_index)
    elif method == 'mean':
        col_list=[0]
        for col in numeric_columns:
            col_list[0] = col
            df_col = outlier_identifier(dataframe, columns=col_list, identifier=identifier, return_df = True)
            outlier_index = df_col[df_col['outlier']==True].index
            dataframe_copy.loc[outlier_index, col] = round(np.mean(dataframe[col]), 2)
    elif method == 'median':
        col_list=[0]
        for col in numeric_columns:
            col_list[0] = col
            df_col = outlier_identifier(dataframe, columns=col_list, identifier=identifier, return_df = True)
            outlier_index = df_col[df_col['outlier']==True].index
            dataframe_copy.loc[outlier_index, col] = round(np.median(dataframe[col]), 2)
    return dataframe_copy

src/outliers/outliers.py
import pandas as pd
import numpy as np
def outlier_identifier(dataframe, columns=None, identifier = 'IQR', return_df=False):
    """
    A function that identify and summarize the count and range of based on the method the user choose
    Parameters
    ----------
    dataframe : pandas.core.frame.DataFrame
        The target dataframe where the function is performed.
    columns : list, default=None
        The target columns where the function needed to be performed. Default is None, the function will check all columns
    identifier : string, default='IQR'
        The method of identifying outliers.
        - if "Z_score" : Use z-test with threshold of 3
        - if "IQR" : Use IQR (Inter Quantile range) to identify outliers (default)
    return_df : bool, default=False
        Can be set to True if want output as dataframe identified with outliers in rows

    Returns
    -------
    pandas.core.frame.DataFrame 
    (a dataframe with the summary of the outlier identified by the method) if return_df = False , 
    (a dataframe with additional column having if row has outlier or not) if return_df = True
    
    Examples
    --------
    >>> import pandas as pd
        
    >>> df = pd.DataFrame({
    >>>    'SepalLengthCm' : [5.1, 4.9, 4.7, 5.5, 5.1, 50, 54, 5.0, 5.2, 5.3, 5.1],
    >>>    'SepalWidthCm' : [1.4, 1.4, 20, 2.0, 0.7, 1.6, 1.2, 1.4, 1.8, 1.5, 2.1],
    >>>    'PetalWidthCm' : [0.2, 0.2, 0.2, 0.3, 0.4, 0.5, 0.5, 0.6, 0.4, 0.2, 5]
    >>> })
    >>> outlier_identifier(df)
    	                SepalLengthCm SepalWidthCm PetalWidthCm
    outlier_count                  2            1            1
    outlier_percentage        18.18%        9.09%        9.09%
    mean                       13.63         3.19         0.77
    median                       5.1          1.5          0.4
    std                        18.99         5.59         1.41
    lower_range                  NaN          NaN          NaN
    upper_range         (50.0, 54.0)         20.0          5.0
    """

    # Handle dataframe type error (Check if dataframe is of type Pandas DataFrame)
    if not isinstance(dataframe, pd.DataFrame):
        raise TypeError(f"passed dataframe is of type {type(dataframe).__name__}, should be DataFrame")
        
    # Handle empty dataframe or dataframe with all NAN
    if dataframe.empty or dataframe.dropna().empty:
        raise ValueError("passed dataframe is None")
    
    # Handle columns type error (Check if columns are None or type list)
    if not columns is None and not isinstance(columns, list):
        raise TypeError(f"passed columns is of type {type(columns).__name__}, should be list or NoneType")
    
    # Handle identifier type error (Check if identifier is of type str)
    if not isinstance(identifier, str):
        raise TypeError(f"passed identifier is of type {type(identifier).__name__}, should be string with value 'Z_score' or 'IQR'")
    
    # Handle identifier Value error (Check if identifier has value 'Z_score' or 'IQR')
    if identifier!='Z_score' and identifier!='IQR':
        raise ValueError("passed identifier should have value 'Z_score' or 'IQR'")
        
    # Handle return_df type error (Check if identifier is of type bool)
    if not isinstance(return_df, bool):
        raise TypeError(f"passed return_df is of type {type(return_df).__name__}, should be bool with value as True or False")

    if columns is None:
        columns = df.columns
    df_filtered = dataframe[columns]
    numeric_columns = df_filtered.select_dtypes('number').columns
    df_selected = df_filtered[numeric_columns]
    df_selected['outlier'] = False
    
    output = pd.DataFrame(columns=numeric_columns, index=['outlier_count', 'outlier_percentage', 'mean', 'median', 'std', 'lower_range', 'upper_range'])
    if identifier == 'Z_score':
        if return_df == False:
            for col in numeric_columns:
                std = df_selected[col].std()
                mean = df_selected[col].mean()
                col_outliers = df_selected[np.abs(df_selected[col] - mean) > std*3]
                output.loc['outlier_count', col] = len(col_outliers)
                output.loc['outlier_percentage', col] = str(round(len(col_outliers) * 100 / len(df_selected), 2)) + '%'
                output.loc['mean', col] = round(mean, 2)
                output.loc['median', col] = round(np.percentile(df_selected[col], 50), 2)
                output.loc['std', col] = round(std, 2)

                low_range = col_outliers[col_outliers[col] < -std*3]
                if not low_range.empty:
                    if len(low_range) == 1:
                        output.loc['lower_range', col] = min(low_range[col])
                    else:
                        output.loc['lower_range', col] = (min(low_range[col]), max(low_range[col]))
                upper_range = col_outliers[col_outliers[col] > std*3]
                if not upper_range.empty:
                    if len(upper_range) == 1:
                        output.loc['upper_range', col] = min(upper_range[col])
                    else:
                        output.loc['upper_range', col] = (min(upper_range[col]), max(upper_range[col]))
            return output
        else:
            for col in numeric_columns:
                std = df_selected[col].std()
                mean = df_selected[col].mean()
                df_selected.loc[np.abs((df_selected[col] - mean)) > std*3, 'outlier'] = True
            return df_selected


    elif identifier == 'IQR':
        if return_df == False:
            for col in numeric_columns:
                std = df_selected[col].std()
                mean = df_selected[col].mean()
                iqr = np.percentile(df_selected[col], 75) - np.percentile(df_selected[col], 25)
                col_outliers = df_selected[np.abs((df_selected[col] - mean) / std) > 1.5*iqr]
                output.loc['outlier_count', col] = len(col_outliers)
                output.loc['outlier_percentage', col] = str(round(len(col_outliers) * 100 / len(df_selected), 2)) + '%'
                output.loc['mean', col] = round(mean, 2)
                output.loc['median', col] = round(np.percentile(df_selected[col], 50), 2)
                output.loc['std', col] = round(std, 2)
                low_range = col_outliers[col_outliers[col] < -1.5*iqr]
                if not low_range.empty:
                    if len(low_range) == 1:
                        output.loc['lower_range', col] = min(low_range[col])
                    else:
                        output.loc['lower_range', col] = (min(low_range[col]), max(low_range[col]))
                upper_range = col_outliers[col_outliers[col] > 1.5*iqr]
                if not upper_range.empty:
                    if len(upper_range) == 1:
                        output.loc['upper_range', col] = min(upper_range[col])
                    else:
                        output.loc['upper_range', col] = (min(upper_range[col]), max(upper_range[col]))
            return output

        else:
            for col in numeric_columns:
                std = df_selected[col].std()
                mean = df_selected[col].mean()
                iqr = np.percentile(df_selected[col], 75) - np.percentile(df_selected[col], 25)
                df_selected.loc[np.abs((df_selected[col] - mean) / std) > 1.5*iqr, 'outlier'] = True
            
            return df_selected